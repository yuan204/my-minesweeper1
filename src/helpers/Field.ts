export enum CellState {
  empty = 0,
  bomb = 9,
  hidden = 10,
  mark = 11,
  weakMark = 12,
}

export type Cell = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;

export type Field = Cell[][];

export type Coord = [Cell, Cell];

export const emptyFieldGenerator = (
  size: number,
  state = CellState.empty
): Field => {
  return Array(size)
    .fill([])
    .map((_) => Array(size).fill(state));
};

export const fieldGenerator = (size: number, probability: number): Field => {
  if (probability < 0 || probability > 1) {
    throw new Error("probability should be between 0 and 1");
  }
  const fields = emptyFieldGenerator(size);
  let restBombCount = size * size * probability;
  let unProcessCount = size * size;
  for (let y = 0; y < size; y++) {
    for (let x = 0; x < size; x++) {
      if (restBombCount === 0) return fields;
      if (restBombCount / unProcessCount > Math.random()) {
        fields[y][x] = CellState.bomb;
        restBombCount--;
      }
      unProcessCount--;
    }
  }
  return fields;
};
