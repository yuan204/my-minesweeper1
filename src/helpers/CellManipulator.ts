import { Coord, Field, CellState } from "./Field";

export const checkItemInField = ([y, x]: Coord, field: Field): boolean => {
  const size = field.length;
  return x >= 0 && y >= 0 && x < size && y < size;
};

export const getNeibours = ([y, x]: Coord): Record<string, Coord> => ({
  left: [y - 1, x],
  right: [y + 1, x],
  top: [y, x - 1],
  bottom: [y, x + 1],
  leftTop: [y - 1, x - 1],
  leftBottom: [y - 1, x + 1],
  rightTop: [y + 1, x - 1],
  rightBottom: [y + 1, x + 1],
});

export const incrementNeibours = (coord: Coord, field: Field): Field => {
  const neibours = getNeibours(coord);
  for (const neibour of Object.values(neibours)) {
    const [y, x] = neibour;
    if (!checkItemInField(neibour, field) || field[y][x] >= CellState.bomb)
      continue;
    field[y][x]++;
  }
  return field;
};

export const openCell = (
  coord: Coord,
  playerField: Field,
  gameField: Field
): Field => {
  const { bomb, empty, hidden } = CellState;

  const [y, x] = coord;
  const gameCell = gameField[y][x];
  if (gameCell === bomb) {
    throw new Error("game over");
  }
  if (gameCell === empty) {
    playerField[y][x] = gameCell;
    const neibours = getNeibours(coord);
    for (const coord of Object.values(neibours)) {
      if (checkItemInField(coord, gameField)) {
        const [y, x] = coord;
        const playerCell = playerField[y][x];
        const gameCell = gameField[y][x];
        if (playerCell === hidden && gameCell !== bomb) {
          openCell(coord, playerField, gameField);
        }
      }
    }
  }
  playerField[y][x] = gameCell;
  return playerField;
};
