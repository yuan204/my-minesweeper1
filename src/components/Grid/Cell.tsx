import React, { FC, MouseEvent } from "react";
import styled from "@emotion/styled";
import { Cell as CellType, CellState, Coord } from "@/helpers/Field";
import { useMouseDown } from "@/hooks/useMouseDown";

export interface CellProps {
  coords: Coord;
  onContextMenu: (coords: Coord) => void;
  onClick: (coords: Coord) => void;
  children: CellType;
}

export const Cell: FC<CellProps> = (props) => {
  const activeCells = [CellState.hidden, CellState.mark, CellState.weakMark];
  const { coords, children } = props;
  const onContextMenu = (event: React.MouseEvent<HTMLElement>) => {
    event.preventDefault();
    if (activeCells.includes(children)) {
      props.onContextMenu(coords);
    }
  };
  const onClick = (event: React.MouseEvent<HTMLDivElement>) =>
    props.onClick(props.coords);
  const cellProps = {
    "data-testid": `${children}-${coords}`,
    children,
    onContextMenu,
    onClick,
    role: "cell",
  };
  return <CellMap {...cellProps} />;
};

interface CellMapProps {
  onContextMenu: (event: React.MouseEvent<HTMLDivElement>) => void;
  onClick: (event: React.MouseEvent<HTMLDivElement>) => void;
  "data-testid": string;
}

export const CellMap: FC<CellMapProps> = ({ children, ...rest }) => {
  const [mousedown, onMouseDown, onMouseUp] = useMouseDown();
  const notActiveProps = {
    onContextMenu: rest.onContextMenu,
    "data-testid": rest["data-testid"],
  };
  const activeProps = {
    ...rest,
    mousedown,
    onMouseDown,
    onMouseUp,
  };
  switch (children) {
    case CellState.hidden:
      return <CloseFrame {...activeProps} />;
    case CellState.empty:
      return <RevealFrame {...notActiveProps} />;
    case CellState.bomb:
      return (
        <BombFrame {...notActiveProps}>
          <Bomb />
        </BombFrame>
      );
    case CellState.mark:
      return (
        <CloseFrame {...activeProps}>
          <Flag />
        </CloseFrame>
      );
    case CellState.weakMark:
      return (
        <CloseFrame {...activeProps}>
          <WeakFlag />
        </CloseFrame>
      );
    default:
      return <RevealFrame {...notActiveProps}>{children}</RevealFrame>;
  }
};

interface CloseFrameProps {
  mousedown?: boolean;
}

export const CloseFrame = styled.div<CloseFrameProps>`
  display: flex;
  align-items: center;
  justify-content: center;
  user-select: none;
  cursor: pointer;
  width: 1.8vw;
  height: 1.8vw;
  background-color: #d1d1d1;
  border: 0.6vh solid transparent;
  border-color: ${(props) =>
    props.mousedown ? "transparent" : "white #9e9e9e #9e9e9e white"};

  &:hover {
    filter: brightness(1.1);
  }
`;
const transparent = "rgba(0,0,0,0)";

const colors: { [key in CellType]: string } = {
  0: "#000000",
  1: "#2a48ec",
  2: "#2bb13d",
  3: "#ec6561",
  4: "#233db7",
  5: "#a6070f",
  6: "#e400af",
  7: "#906a02",
  8: "#fa0707",
  9: transparent,
  10: transparent,
  11: transparent,
  12: transparent,
};
const RevealFrame = styled(CloseFrame)`
  border-color: #dddddd;
  cursor: default;

  &:hover {
    filter: brightness(1);
  }

  color: ${({ children }) => colors[children as CellType]};
`;

const BombFrame = styled(RevealFrame)`
  background: red;
`;

const Bomb = styled.div`
  border-radius: 50%;
  width: 1vw;
  height: 1vw;
  background-color: #333;
`;

const Flag = styled.div`
  width: 0px;
  height: 0px;
  border-top: 0.5vw solid transparent;
  border-bottom: 0.5vw solid transparent;
  border-left: 0.5vw solid #ec433c;
`;

const WeakFlag = styled(Flag)`
  border-left: 0.5vw solid #f19996;
`;
