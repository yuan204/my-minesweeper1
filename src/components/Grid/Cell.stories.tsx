import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import { Cell } from "./Cell";

export default {
  title: "Top/Cell",
  component: Cell,
  argTypes: {
    coords: {
      defaultValue: [2, 2],
    },
  },
} as ComponentMeta<typeof Cell>;

const Template: ComponentStory<typeof Cell> = (args) => <Cell {...args} />;

export const CellEmpty = Template.bind({});
CellEmpty.args = {
  children: 0,
};

export const CellHidden = Template.bind({});
CellHidden.args = {
  children: 10,
};

export const CellBomb = Template.bind({});
CellBomb.args = {
  children: 9,
};

export const CellWeakMark = Template.bind({});
CellWeakMark.args = {
  children: 12,
};

export const CellMark = Template.bind({});
CellMark.args = {
  children: 11,
};

export const CellWith1 = Template.bind({});
CellWith1.args = {
  children: 1,
};

export const CellWith2 = Template.bind({});
CellWith2.args = {
  children: 2,
};

export const CellWith3 = Template.bind({});
CellWith3.args = {
  children: 3,
};

export const CellWith8 = Template.bind({});
CellWith8.args = {
  children: 8,
};
