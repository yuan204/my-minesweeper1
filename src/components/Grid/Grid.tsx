import React, { FC } from "react";
import { Coord, Field } from "@/helpers/Field";
import { Cell } from "@/components/Grid/Cell";
import styled from "@emotion/styled";

export interface GridProps {
  children: Field;
  onContextMenu: (coords: Coord) => void;
  onClick: (coords: Coord) => void;
}
export const Grid: FC<GridProps> = ({ children, ...rest }) => (
  <Wrapper size={children.length}>
    {children.map((row, y) =>
      row.map((item, x) => {
        const coords: Coord = [y, x];
        const cellProps = {
          children: item,
          coords,
          ...rest,
        };
        return <Cell {...cellProps} key={`${y}-${x}`} />;
      })
    )}
  </Wrapper>
);
interface WrapperProps {
  size: number;
}
const Wrapper = styled.div<WrapperProps>`
  display: grid;
  grid-template-columns: repeat(${({ size }) => size}, auto);
  width: max-content;
  padding: 1vw;
`;
