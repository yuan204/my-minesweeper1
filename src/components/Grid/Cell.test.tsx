import { CellState, Coord } from "@/helpers/Field";
import { fireEvent, render, screen, createEvent } from "@testing-library/react";
import { Cell, CellProps, CloseFrame } from "@/components/Grid/Cell";
import React from "react";

describe("cell check", () => {
  const activeCells = [CellState.hidden, CellState.mark, CellState.weakMark];
  const coords: Coord = [1, 1];
  for (let i = CellState.empty; i <= CellState.weakMark; i++) {
    const onClick = jest.fn();
    const onContextMenu = jest.fn();
    const cellProps: CellProps = {
      coords,
      children: i,
      onClick: onClick,
      onContextMenu: onContextMenu,
    };
    test("cell should render correctly", () => {
      const { asFragment } = render(<Cell {...cellProps} />);
      expect(asFragment()).toMatchSnapshot();
    });
    test("closeFramed should render correctly", () => {
      const { asFragment } = render(<CloseFrame mousedown={true} />);
      expect(asFragment()).toMatchSnapshot();
    });
    // it  unmount React tree after each test auto called cleanup
    test("should prevent default contextMenu for every type cell", () => {
      render(<Cell {...cellProps} />);
      const element = screen.getByTestId(`${i}-${coords}`);
      const contextMenuEvent = createEvent.contextMenu(element);
      fireEvent(element, contextMenuEvent);
      expect(contextMenuEvent.defaultPrevented).toBe(true);
    });
    test("onclick and onContextMenu should be called for active cells", () => {
      render(<Cell {...cellProps} />);
      const element = screen.getByTestId(`${i}-${coords}`);
      fireEvent.contextMenu(element);
      fireEvent.click(element);
      if (activeCells.includes(i)) {
        expect(onContextMenu).toBeCalled();
        expect(onClick).toBeCalled();
      } else {
        expect(onContextMenu).not.toBeCalled();
        expect(onClick).not.toBeCalled();
      }
    });
  }
});
