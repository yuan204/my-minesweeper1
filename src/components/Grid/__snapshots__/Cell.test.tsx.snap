// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cell check cell should render correctly 1`] = `
<DocumentFragment>
  <div
    class="css-oq4emy"
    data-testid="0-1,1"
  />
</DocumentFragment>
`;

exports[`cell check cell should render correctly 2`] = `
<DocumentFragment>
  <div
    class="css-1pfp8zg"
    data-testid="1-1,1"
  >
    1
  </div>
</DocumentFragment>
`;

exports[`cell check cell should render correctly 3`] = `
<DocumentFragment>
  <div
    class="css-spzl2v"
    data-testid="2-1,1"
  >
    2
  </div>
</DocumentFragment>
`;

exports[`cell check cell should render correctly 4`] = `
<DocumentFragment>
  <div
    class="css-qwh4sq"
    data-testid="3-1,1"
  >
    3
  </div>
</DocumentFragment>
`;

exports[`cell check cell should render correctly 5`] = `
<DocumentFragment>
  <div
    class="css-31jsrs"
    data-testid="4-1,1"
  >
    4
  </div>
</DocumentFragment>
`;

exports[`cell check cell should render correctly 6`] = `
<DocumentFragment>
  <div
    class="css-1fzczfp"
    data-testid="5-1,1"
  >
    5
  </div>
</DocumentFragment>
`;

exports[`cell check cell should render correctly 7`] = `
<DocumentFragment>
  <div
    class="css-e77skh"
    data-testid="6-1,1"
  >
    6
  </div>
</DocumentFragment>
`;

exports[`cell check cell should render correctly 8`] = `
<DocumentFragment>
  <div
    class="css-aljgbs"
    data-testid="7-1,1"
  >
    7
  </div>
</DocumentFragment>
`;

exports[`cell check cell should render correctly 9`] = `
<DocumentFragment>
  <div
    class="css-n61ljn"
    data-testid="8-1,1"
  >
    8
  </div>
</DocumentFragment>
`;

exports[`cell check cell should render correctly 10`] = `
<DocumentFragment>
  <div
    class="css-1xn280i"
    data-testid="9-1,1"
  >
    <div
      class="css-3tubcw"
    />
  </div>
</DocumentFragment>
`;

exports[`cell check cell should render correctly 11`] = `
<DocumentFragment>
  <div
    class="css-5803oy"
    data-testid="10-1,1"
  />
</DocumentFragment>
`;

exports[`cell check cell should render correctly 12`] = `
<DocumentFragment>
  <div
    class="css-5803oy"
    data-testid="11-1,1"
  >
    <div
      class="css-1w1u3hh"
    />
  </div>
</DocumentFragment>
`;

exports[`cell check cell should render correctly 13`] = `
<DocumentFragment>
  <div
    class="css-5803oy"
    data-testid="12-1,1"
  >
    <div
      class="css-1wbzzpj"
    />
  </div>
</DocumentFragment>
`;

exports[`cell check closeFramed should render correctly 1`] = `
<DocumentFragment>
  <div
    class="css-1mjcmhi"
  />
</DocumentFragment>
`;

exports[`cell check closeFramed should render correctly 2`] = `
<DocumentFragment>
  <div
    class="css-1mjcmhi"
  />
</DocumentFragment>
`;

exports[`cell check closeFramed should render correctly 3`] = `
<DocumentFragment>
  <div
    class="css-1mjcmhi"
  />
</DocumentFragment>
`;

exports[`cell check closeFramed should render correctly 4`] = `
<DocumentFragment>
  <div
    class="css-1mjcmhi"
  />
</DocumentFragment>
`;

exports[`cell check closeFramed should render correctly 5`] = `
<DocumentFragment>
  <div
    class="css-1mjcmhi"
  />
</DocumentFragment>
`;

exports[`cell check closeFramed should render correctly 6`] = `
<DocumentFragment>
  <div
    class="css-1mjcmhi"
  />
</DocumentFragment>
`;

exports[`cell check closeFramed should render correctly 7`] = `
<DocumentFragment>
  <div
    class="css-1mjcmhi"
  />
</DocumentFragment>
`;

exports[`cell check closeFramed should render correctly 8`] = `
<DocumentFragment>
  <div
    class="css-1mjcmhi"
  />
</DocumentFragment>
`;

exports[`cell check closeFramed should render correctly 9`] = `
<DocumentFragment>
  <div
    class="css-1mjcmhi"
  />
</DocumentFragment>
`;

exports[`cell check closeFramed should render correctly 10`] = `
<DocumentFragment>
  <div
    class="css-1mjcmhi"
  />
</DocumentFragment>
`;

exports[`cell check closeFramed should render correctly 11`] = `
<DocumentFragment>
  <div
    class="css-1mjcmhi"
  />
</DocumentFragment>
`;

exports[`cell check closeFramed should render correctly 12`] = `
<DocumentFragment>
  <div
    class="css-1mjcmhi"
  />
</DocumentFragment>
`;

exports[`cell check closeFramed should render correctly 13`] = `
<DocumentFragment>
  <div
    class="css-1mjcmhi"
  />
</DocumentFragment>
`;
